src/yb/rocksdb/db/db_impl.cc

```diff
@@ -3285,19 +3285,23 @@ void DBImpl::SchedulePendingCompaction(ColumnFamilyData* cfd) {
   TEST_SYNC_POINT("DBImpl::SchedulePendingCompaction:Done");
 }
 
-void DBImpl::BGWorkFlush(void* db) {
+void DBImpl::BGWorkFlush(void* db, yb::util::WaitStateInfoPtr wait_state) {
```

Comment:
Where are we actually using this call, with 2 params? I can't find any caller for this.

Also wondering why we need to pass in the parameter and set the status here -- wouldn't the same be achieved by setting Wait_state at the start/end of BackgroundCallFlush method?

Similarly for BackgroundCallCompaction.

Reply:
The caller is in thread_posix.cc line 136.

Since there are multiple threads which can do flush and compaction, a single wait_state object won't be enough. We need a wait_state object per thread, and a way to access the wait states of all the threads. I found the best place for that to be in the class ThreadPool (in thread_posix.h).

There are 2 ways to have this wait_state object in the thread pool, one is to go from BGWorkFlush to the thread pool to create the wait state object, this won't change the function signature, but will require locks when we are creating and destroying this wait state object. And we have to create and destroy wait state object every time flush and compaction runs.

The other way is to create one wait state object per thread, that way we don't have to acquire any extra locks, but doing this requires us to somehow pass this wait state object from the thread pool to BGWorkFlush, that's why the function signature changes. After completion of the job, we just have to update the status to unused.

Reply:
We'll probably need some kind of locking ... to ensure that DumpRpcs get the latest state. But, keeping that aside, even with having the vector, this seems like something that can be avoided.

Before calling
```
133: (*function)(arg);
```
we can call the appropriate `ADOPT_WAIT_STATE(...);`

and the functions in Compact/Flush can update the "ThreadLocal" wait state using SET_WAIT_STATE in the compaction/flush function calls.